examples/tpm2-seal-unseal/main.go

```diff
@@ -0,0 +1,281 @@
+		errs = append(errs, fmt.Errorf("error while unsealing data: %v", otherErr))
+	}
+	if len(errs) > 0 {
+		return
+	}
+	fmt.Printf("Unsealed data: 0x%x\n", unsealedData)
+
+	// Try to unseal the data with the wrong password
+	_, unsealErr, otherErrs = unseal(rwc, pcr, "wrong-password", objectHandle)
+	if unsealErr == nil && len(otherErrs) == 0 {
+		errs = append(errs, fmt.Errorf("expected unsealing with wrong password to fail but it succeeded"))
+	}
+	for otherErr := range otherErrs {
+		errs = append(errs, fmt.Errorf("error while unsealing data: %v", otherErr))
+	}
+	if len(errs) > 0 {
+		return
+	}
+	// Note that this can be easily fooled since we don't verify that the error was
+	// due specifically to policy.
+	fmt.Printf("As expected, unsealing with the wrong password failed with an error. Error was: %v\n", unsealErr)
+
+	// Extend the PCR
+	if err = tpm2.PCREvent(rwc, tpmutil.Handle(pcr), []byte{1}); err != nil {
+		return append(errs, fmt.Errorf("unable to extend PCR: %v", err))
+	}
+	fmt.Printf("Extended PCR %d\n", pcr)
+
+	// Note the new value of the pcr
+	pcrVal, err = readPCR(rwc, pcr)
+	if err != nil {
+		return append(errs, fmt.Errorf("unable to read PCR: %v", err))
+	}
+	fmt.Printf("PCR %d value: 0x%x\n", pcr, pcrVal)
+
+	// Try to unseal the data with the PCR in the wrong state
+	_, unsealErr, otherErrs = unseal(rwc, pcr, objectPassword, objectHandle)
+	if unsealErr == nil && len(otherErrs) == 0 {
+		errs = append(errs, fmt.Errorf("expected unsealing with wrong PCR state to fail but it succeeded"))
+	}
+	for otherErr := range otherErrs {
+		errs = append(errs, fmt.Errorf("error while unsealing data: %v", otherErr))
+	}
+	if len(errs) > 0 {
+		return
+	}
+	// Note that this can be easily fooled since we don't verify that the error was
+	// due specifically to policy.
+	fmt.Printf("As expected, unsealing after extending the PCR failed with an error. Error was: %v\n", unsealErr)
+
+	return
+}
+
+// Returns the unsealed data, any error due to the Unseal command, and any other errors
+func unseal(rwc io.ReadWriteCloser, pcr int, objectPassword string, objectHandle tpmutil.Handle) (data []byte, unsealErr error, otherErrs []error) {
+	// Create the authorization session
+	sessHandle, _, err := policyPCRPasswordSession(rwc, pcr, objectPassword)
+	defer func() {
+		if sessHandle != tpm2.HandleNull {
+			if flushErr := tpm2.FlushContext(rwc, sessHandle); flushErr != nil {
+				otherErrs = append(otherErrs, fmt.Errorf("unable to flush session: %v", flushErr))
+			}
+		}
+	}()
+	if err != nil {
+		otherErrs = append(otherErrs, fmt.Errorf("unable to get auth session: %v", err))
+		return
+	}
+
+	// Unseal the data
+	unsealedData, err := tpm2.UnsealWithSession(rwc, sessHandle, objectHandle, objectPassword)
+	if err != nil {
+		unsealErr = fmt.Errorf("unable to unseal data: %v", err)
+		return
+	}
+	data = unsealedData
+	return
+}
+
+// Note that the handle may be returned even if an error is returned as well, in which
```

Comment:
This func should handle it internally. Something like:

```
tpm2.StartAuthSession(...)
// check err
defer func() {
  // clean up the handle if we had an error
  if err != nil {
    tpm2.FlushContext(sessHandle)
  }
} ()
```

Reply:
Having this method try to clean up that way complicates things since it needs to report whether the flushing failed or not which leads to it returning multiple errors...WDYT?

Reply:
It's just kicking the problem up the stack.
You also don't have to return multiple errors:

```
return fmt.Errorf("%v\nfailed flushing auth session handle: %v", err, flushErr)
```

Reply:
will change this